File: alloc_min.cpp
#include <cstdlib>
#include <chrono>
#include <iostream>
#include <list>
#include <vector>
#include <cstring>

#define HEAP_PAGE_SIZE_WORDS 128000
#define SMALL_ALLOC_WORDS (32 + 1)
#define HEAP_PAGE_MIN_FREE (HEAP_PAGE_SIZE_WORDS - SMALL_ALLOC_WORDS)

class Allocator {
    std::list<uintptr_t*> pages;
    uintptr_t *lastPage;
    long currPageWordsUsed;

    void add_page(unsigned long size = HEAP_PAGE_SIZE_WORDS) {
        this->pages.push_back(new uintptr_t[size]);
        this->lastPage = this->pages.back();
        this->currPageWordsUsed = 0;
    }

public:
    explicit Allocator() {
        this->add_page();
    }

    void* alloc_small(uintptr_t type, long size) {
        if (currPageWordsUsed > HEAP_PAGE_MIN_FREE) {
            this->add_page();
        }

        lastPage[currPageWordsUsed] = (uintptr_t) lastPage;
        lastPage[currPageWordsUsed + 1] = type;
        uintptr_t* alloc = lastPage + currPageWordsUsed + 2;
        currPageWordsUsed += 2 + size;

        std::memset(alloc, 0, size); // faster than the libsodium method

        return alloc;
    }

    void *alloc_big(uintptr_t type, long size) {
        if (HEAP_PAGE_SIZE_WORDS - 2 - this->currPageWordsUsed >= size)  {
        } else {
            this->add_page();
        }

        lastPage[currPageWordsUsed] = (uintptr_t) lastPage;
        lastPage[currPageWordsUsed + 1] = type;
        uintptr_t* alloc = lastPage + currPageWordsUsed + 2;
        currPageWordsUsed += 2 + size;

        std::memset(alloc, 0, size);

        return alloc;
    }

    void wipe() {
        for (auto *page : this->pages) {
            delete[] page;
        }
        this->pages.clear();
        this->add_page();
    }
};

const auto size = 1024;
const long iters = 1024 * 256 * 3;

int main() {
    Allocator alloc = Allocator();

    auto **nodes = new uintptr_t*[iters];

    auto start = std::chrono::high_resolution_clock::now();

    for (long iter = 0; iter < size; iter++) {
        for (int i = 0; i < iters; i++) {
//            nodes[i] =
                    (uintptr_t* ) alloc.alloc_small(1, 2);
        }
        if (iter % 2 == 0) alloc.wipe();
    }

    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - start).count();

    std::cout << ms << std::endl;

    std::cout << "Allocs per sec (in millions) " << ((size * iters) / ms * 1000) / 1000000.0 << std::endl;
}

File: alloc_min.rs
mod memset;

use std::alloc::{alloc, Layout};
use std::collections::LinkedList;
use std::env;
use std::mem::{align_of, size_of};
use std::ptr::{slice_from_raw_parts_mut, write_volatile};

const HEAP_PAGE_SIZE_WORDS: usize = 128000;

type WORD = usize;

unsafe fn layout(size: WORD) -> Layout {
    Layout::from_size_align_unchecked(size_of::<WORD>() * size, align_of::<WORD>())
}


struct Allocator {
    pages: LinkedList<*mut WORD>,
    curr_page_words_used: WORD,
}

impl Allocator {

    unsafe fn new() -> Allocator {
        let mut allocator = Allocator {
            pages: Default::default(),
            curr_page_words_used: 0,
        };

        allocator.add_page(HEAP_PAGE_SIZE_WORDS);

        allocator
    }

    unsafe fn add_page(&mut self, size: WORD) {
        let page_ptr: *mut WORD = std::mem::transmute(std::alloc::alloc(layout(size)));
        self.pages.push_back(page_ptr);

        // it seems performance beneficial to write to the newly allocated page
        // even if the value is not used later
        *page_ptr = size;

        if cfg!(feature = "page-size-tracking") {
            self.curr_page_words_used = 1;
        } else {
            self.curr_page_words_used = 0;
        }
    }

    unsafe fn alloc<const TYP: WORD, const SIZE: WORD>(&mut self) -> *const WORD {
        if self.curr_page_words_used > HEAP_PAGE_SIZE_WORDS - SIZE - 2 {
            self.add_page(HEAP_PAGE_SIZE_WORDS)
        }

        let last_page = *self.pages.back_mut().unwrap();

        *last_page.add(self.curr_page_words_used) = std::mem::transmute(last_page);
        *last_page.add(self.curr_page_words_used + 1) = TYP;

        let alloc = last_page.add(self.curr_page_words_used + 2);

        for offset in 0..SIZE {
            alloc.add(offset).write(0);
        }

        self.curr_page_words_used += 2 + SIZE;

        alloc
    }

    unsafe fn wipe(&mut self) {
        for page in self.pages.iter() {
            let size = if cfg!(feature = "page-size-tracking") {
                **page as usize
            } else {
                0
            };
            std::alloc::dealloc(std::mem::transmute(*page), layout(size));
        }
        self.pages.clear();
        self.add_page(HEAP_PAGE_SIZE_WORDS);
    }
}

const SIZE: WORD = 1024;
const ITERS: WORD = 1024 * 256 * 3;

fn main() {
    let size = env::args().len() * 2;

    unsafe {
        let mut alloc = Allocator::new();

        let start = std::time::Instant::now();

        for iter in 0..SIZE {
            for _i in 0..ITERS {
                alloc.alloc::<1, 0>();
            }
            if iter % 1 == 0 {
                alloc.wipe();
            }
        }

        let end = std::time::Instant::now();

        let ms = end.checked_duration_since(start).unwrap().as_millis() as WORD;

        println!("{}", ms);
        println!("Allocs per sec (in millions) {}", ((SIZE * ITERS) / ms * 1000) / 1000000);
    }
}
